controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,418 @@
+	constraintName := labels["internal.gatekeeper.sh/constraint-name"]
+
+	syncOnlySet := map[v1alpha1.SyncOnlyEntry]bool{}
+	// Add to table for unique filtering
+	for _, entry := range config.Spec.Sync.SyncOnly {
+		syncOnlySet[entry] = true
+	}
+	// start
+	constraintGVR := schema.GroupVersionResource{
+		Group:    "constraints.gatekeeper.sh",
+		Version:  "v1beta1",
+		Resource: strings.ToLower(constraintKind),
+	}
+
+	constraint, err := r.DynamicClient.Resource(constraintGVR).Get(ctx, constraintName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("The Constraint is deleted in the config resouce", "constraintName:", constraint.GetName())
+
+			return reconcile.Result{}, nil //nolint:nilerr
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	constraintMatchKinds, _, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
+	if err != nil {
+		r.Log.V(1).Info("There is no provided kinds in contsraint", "constraintName:", constraint.GetName())
+
+		return reconcile.Result{}, nil //nolint:nilerr
+	}
+	// remove
+	constraintSyncOnlyEntries, err := r.getSyncOnlys(constraintMatchKinds)
+	if err != nil {
+		if errors.Is(err, ErrNotFoundDiscovery) {
+			return reconcile.Result{RequeueAfter: time.Minute * 3}, nil
+		}
+
+		log.Error(err, "Error to get matching kind and apigroup")
+
+		return reconcile.Result{}, nil
+	}
+
+	r.addConstraintToMap(constraint.GetName(), *constraintSyncOnlyEntries)
+	updatedSyncOnly := r.getUniqSyncOnly()
+	err = r.updateConfigSyncOnlyEntry(ctx, config, updatedSyncOnly)
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ConstraintStatusReconciler) getSyncOnlys(constraintMatchKinds []interface{}) (
+	*[]v1alpha1.SyncOnlyEntry, error,
+) {
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+
+	for _, match := range constraintMatchKinds {
+		newKind, ok := match.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		apiGroups, ok := newKind["apiGroups"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		kindsInKinds, ok := newKind["kinds"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		for _, apiGroup := range apiGroups {
+			for _, kind := range kindsInKinds {
+				version, err := r.getAPIVersion(kind.(string), apiGroup.(string), false)
+				if err != nil {
+					return nil, err
+				}
+
+				syncOnlys = append(syncOnlys, v1alpha1.SyncOnlyEntry{
+					Group:   apiGroup.(string),
+					Version: version,
+					Kind:    kind.(string),
+				})
+			}
+		}
+	}
+
+	return &syncOnlys, nil
+}
+
+func (r *ConstraintStatusReconciler) getAPIVersion(kind string, apiGroup string, skipRefresh bool) (string, error) {
+	// cool time(10 min) to refresh discoveries
+	if len(r.apiResourceList) == 0 ||
+		r.discoveryLastRefreshed.Add(time.Minute*10).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		r.discoveryLastRefreshed = time.Now()
+
+		if err != nil {
+			return "", err
+		}
```

Comment:
You don't need to refresh again below if you just refreshed it.
```suggestion
		}

		skipRefresh = true
```